xbindkeys.1
.TH XBindKeys 1 "Tue Apr 19 2014"
.UC 1
.SH "NAME"
xbindkeys \- a grabbing keys program for X
.SH "SYNOPSIS"
\fBxbindkeys\fP [ \fIoptions\fP ]
.br
.SH "DESCRIPTION"
\fIXbindKeys\fP is a program that grabs key and mouse button events
in X and starts associated shell commands.

.LP
.SH "COMMAND LINE OPTIONS"
Available command line options are as follows:
.TP
\fB-d, --defaults\fP
Print a default configuration file
.TP
\fB-f, --file\fP
Use an alternative configuration file
.TP
\fB-h, --help\fP
Short help on options plus version/author info.
.TP
\fB-X, --display\fP
Set X display to use
.TP
\fB-v, --verbose\fP
More information on xbindkeys when it runs

.TP
\fB-k, --key\fP
Identify one key pressed (useful for configuration)

.TP
\fB-m, --multikey\fP
Identify multi key pressed (useful for configuration)

.TP
\fB-g, --geometry\fP
size and position of window with -k|-m option

.TP
\fB-n, --nodaemon\fP
don't start as daemon

.SH "XBINDKEYSRC"
XBindKeys default configuration file is at $HOME/.xbindkeysrc,
used only for one user.
.LP
The configuration file consists of commands in double quotes,
and associated keys in the next line.
.LP
Comments are started with a semicolon (;) and extend to the end of the line.
.LP
You can see a default configuration with --defaults
.LP
.fi

.SH "KEYSYMS"
The program expects combinations of keysyms to be used by using plus (+)
characters to combine them together.
.LP
The format of a command line is:
.nf
    "command to start &"
.nf
       associated key
.LP
.fi
To specify an associated key, you can use `xbindkeys --key` or
 `xbindkeys --multikey` and put one of the two lines in the
configuration file.
.LP
A list of keys is in /usr/include/X11/keysym.h and in
/usr/include/X11/keysymdef.h.
The XK_ is not needed.
.LP
List of modifiers:
.nf
    Release, Control, Shift, Mod1 (Alt), Mod2 (NumLock),
.nf
    Mod3 (CapsLock), Mod4, Mod5 (Scroll).
.LP
.fi
The release modifier is not a standard X modifier, but you can
use it if you want to catch release events instead of press events.
Modifiers are not case sensitive.
.LP
By defaults, xbindkeys does not pay attention to the modifiers
NumLock, CapsLock and ScrollLock.
Add the lines below in the configuration file if you want to pay
attention to them.
.LP
    keystate_numlock = enable
.nf
    keystate_capslock = enable
.nf
    keystate_scrolllock= enable

.SH "EXAMPLES"
.LP
# control+shift+q starts xterm
.nf
"xterm"
.nf
  control+shift + q
.LP
# Menu key starts xbindkeys_show
.nf
"xbindkeys_show"
.nf
  Menu
.LP
# Control + mouse button 1 starts xterm
.nf
"xterm"
.nf
  Control + b:1
.LP
# Control+Shift+a  release event starts rxvt
.nf
"rxvt"
.nf
  release+control+shift + a
.LP
# Control + mouse button 2 release event starts rxvt
.nf
"rxvt"
.nf
  Control + b:2 + Release

.SH "SCHEME CONFIGURATION STYLE"
A scheme configuration can be used if guile support has not been disabled.
The standard configuration is loaded only if the scheme configuration is not
available.
.LP
For more details, please see here
.nf
  http://www.gnu.org/software/guile/guile.html
.LP
\fBNote: The guile configuration file is prefered for non trivial
configurations, for example double click, timed click or key combinations.\fP
.LP
The default location is ~/.xbindkeysrc.scm
.LP
Use xbindkeys --defaults-guile for more details.
.LP
And xbindkeys --defaults-guile > ~/.xbindkeysrc.scm to use this method.

.SH "NOTE"
xbindkeys reloads the configuration file if modified.
Send a HUP signal to force reloading:
.LP
killall -HUP xbindkeys
.LP
When a HUP signal is sent to xbindkeys, changes take effect only after
moving the mouse.

.SH "BUGS"
If you find a bug, please send mail to <xbindkeys-devel@nongnu.org>
.LP
.SH "HOMEPAGE"
.nf
http://www.nongnu.org/xbindkeys/
.fi

.LP
.SH "AUTHOR"
Philippe Brochard <hocwp@free.fr>
.nf
Marcello Mathias Herreshoff for guile support <marcello@hummer.stanford.edu>

.LP
.SH "FILES"
.nf
$HOME/.xbindkeysrc
        The users configuration file.
.LP
$HOME/.xbindkeysrc.scm
        The users configuration file in scheme
        (if guile support has not been disabled).
.fi

.LP
.SH "SEE ALSO"
.nf

\fBxbindkeys_show\fP
    Utility for showing the actual keybinding with xbindkeys
\fBxmodmap(1x)\fP
    Utility for modifying keymap & button mappings in X.
\fBxev(1x)\fP
    Print contents of X events.
\fB/usr/include/X11/keysymdef.h\fP
    X11 KeySym definitions.



ChangeLog
2019-07-19  Zsombor Barna <unibro@mailo.com>
	* options.c: rewrote option parsing to popt
	* all: deleted most global variables
	* all: added parameters to functions which needed globals
	* all: reworked verbose flag
	* options.c: moved functions to xbindkeys.c and new the util.c
	* xbindkeys.c: main function expanded, inner_main shortened
	* xbindkeys.c: added function to do init that needs guile

2019-07-17  Zsombor Barna <unibro@mailo.com>
	* options.c: Removed support of old rc configuration scheme
	* options.c: Rewrote option parsing to use getopt
	* xbindkeys.c: Removed support of old rc configuration scheme
2011-03-09  Hans-Peter Deifel <hpdeifel@gmx.de>

	* options.c: Replace some deprecated guile functions.
	The GH interface has been removed in guile-2.0 so do not use it
	anymore in xbindkeys.

2011-02-21  Philippe Brochard  <pbrochard@common-lisp.net>

	* options.c (get_rc_guile_file): Fix a crash related to guile's GC
	and freeing of static memory. Thanks to Thomas Lindroth.

2010-12-04  Philippe Brochard  <pbrochard@common-lisp.net>

	* options.c: Fix some compilations warnings.

	* xbindkeys.c (adjust_display): adjust envstr_size to prevent a
	segfault with libX11-1.4.0.

2010-01-24  Alexander Clouter <alex@digriz.org.uk>

	* xbindkeys.c (event_loop): Doesn't stat on configuration file
	change the first time if poll-rc is false.

2010-01-23  Philippe Brochard  <hocwp@free.fr>
	* options.c: A poll-rc flag has been added to poll the rc/guile
	configuration file for update. By default no poll is done on
	configuration files. Many thanks to Alexander Clouter.

	* xbindkeys.c (event_loop): Stat both default rc file and guile rc
	file for reloading them when they change.

xbindkeys changelog by Philippe Brochard <hocwp@free.fr>


version:

1.8.3:
	Add a listen loop in the main loop. This fix a bug where, in the
	previous version, a keypress was needed to reload the
	configuration file after a keybinding change. Now one can edit the
	configuration file and keybindings are automatically updated after
	each save. Many thanks to Edwin Stang!

1.8.2:
	Correction of typos in the man page. An emphasis is made on the
	documentation about the guile configuration file (the prefered way
	for non trivial configuration).
	Many thanks to Brad Sawatzky.

1.8.1:
	Better Guile configuration with autotools.
	Minor compilation warning have been fixed.

1.8.0:
	Enable a full access to the xbindkeys internal from the guile
	scheme configuration file. A grabbed key can start a shell command
	or run a scheme function. This enable to do more powerfull things
	inside xbindkeys whitout the need of an external shell script
	(like double click, timed double click or keys combinations).

---------------------------------------------------------------------------

1.7.4:
	xbindkeys.spec: Change Copyright to License.
	Little english correction in warnings.
	Configuration files are reloaded on the fly when they have
	changed so there is no more need to send a HUP signal to reload
	them (Many thanks to Shish).
	Use `scm_c_define_gsubr' instead of 'scm_make_gsubr' with the
	guile mode.


1.7.3:
	Bugfix under Solaris, XBindKeys built with -DFORK_FLAG=1 fails to
	wait for processes it spawns, leaving lots of defunct processes.
	Thanks to Ciaran Anscomb for the patch.

1.7.2:
	Changes to grab keys in all display screens and not only in the
	default root window (Xinerama suport).
	Bugfix for the Guile support in configure.ac.

1.7.1:
	Verbose mode is not started as daemon.
	Check rc file before starting as daemon (so warning about
	configuration files are visible).

1.7.0:
	Add guile/scheme configurtion file style (Many thanks to Marcello
	Mathias Herreshoff).

---------------------------------------------------------------------------

1.6.5:
        Add a --disable-tk option to prevent xbindkeys_show install.
	So no tcl/tk dependency is needed.
	Add a NOTE in xbindkeys man page to say that xbindkeys can
	reread its config file with a -HUP signal (killall -HUP xbindkeys).
	Add two verbose test (forgotten in other version).

1.6.4:
	Add a --disable-fork option  in configure.ac to start one command
	at a time (use system instead of fork+exec).
	Remove umask (022) in daemonize function which cause troubles.

1.6.3:
	Change to switch to autoconf2.5X and automake >= 1.6
	Remove chdir ("/tmp")  in daemonize function.

1.6.2:
	Remove AC_PROG_RANLIB and AC_PROG_MAKE_SET from configure.in
	to prevent rebuit of Makefile when doing a make.

1.6.1:
	Minor buggfix : use strtol instead atoi for use with
	hexadecimal number.
	Complete daemonzie function

1.6.0:
	Add: Release event for keys and buttons.
	Don't use threads any more (use fork+exec instead)
	Better daemonize function
	Simplification of configure.in (better test for X)
	Don't use sscanf any more

---------------------------------------------------------------------------

1.5.5:
	Change: better identification of modifiers with --key and
	--mulitkey options.

1.5.4:
        Change: add warning when a key is unknown in .xbindkeysrc.

1.5.3:
	Bugfix: if the keycode is unknow then xbindkeys don't asign a
	key with keycode=0.
	Change: The default RC file don't use the Menu key witch cause
	problems.

1.5.2:
	Change: if there's a conflict with keys then xbindkeys just
	make a warning and continue with other keys.

1.5.1:
	Add: a new interface for xbindkeys_show (use scrollbar).

1.5.0:
	Add: xbindkeys have now the ability to catch Mouse Button Events.
	so: "xterm"
	      control + b:2
	starts an xterm when mouse button 2 and control are pressed.

---------------------------------------------------------------------------

1.4.1:
	Solve some minor problems when compiling with gcc 3.0.4 under GNU/Linux.

1.4.0:
	Change: rename all code to make it uniform.
	Add: start as daemon by default, but can be changed with
	  --nodaemon|-n option.
	Add: xbindkeys starts commands in threads, so they are not blocking
	  it if we forget the & at the end of the command (this can be

xbindkeys.c
/***************************************************************************
        xbindkeys : a program to bind keys to commands under X11.
                           -------------------
    begin                : Sat Oct 13 14:11:34 CEST 2001
    copyright            : (C) 2001 by Philippe Brochard
    email                : hocwp@free.fr
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/

//#include <config.h>
#include <X11/Xlib.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>
//#include <fcntl.h>

#include "keys.h"

#include "get_key.h"
#include "grab_key.h"
#include "options.h"

#include <libguile.h>

#include "util.h"
#include <X11/XKBlib.h>
#include <popt.h>

static void inner_main(void * /*passed_data*/, int /*argc*/, char ** /*argv*/);
Display *current_display; // The current display

static int got_hup;

char *geom;

void catch_hup_signal(int sig) {
  got_hup = 1;

#ifdef AVOID_KNOWN_HARMLESS_WARNINGS
  sig = sig;
#endif
}

void catch_chld_signal(int sig) {
  pid_t child;

  /*   If more than one child exits at approximately the same time, the signals
   */
  /*   may get merged. Handle this case correctly. */
  while ((child = waitpid(-1, NULL, WNOHANG)) > 0) {
#ifdef DEBUG
    printf("Catch CHLD signal -> pid %i terminated\n", child);
#endif
  }
#ifdef AVOID_KNOWN_HARMLESS_WARNINGS
  sig = sig;
#endif
}

int main(const int ARGC, const char **argv) {

  // guile shouldn't steal our arguments! we already parse them!
  // so we put them in temporary variables.
  got_hup = 0;
  char c;
  char *home;

  char default_file[513];

  home = getenv("HOME");

  strncpy(default_file, home, sizeof(default_file) - 24);
  strncat(default_file, "/.config/xbindkeysrc.scm", sizeof(default_file) - strlen(default_file) - 1);

  // Option definitions
  int verbose = 0;
  char *rc_guile_file = NULL;
  int have_to_show_binding = 0;
  int have_to_get_binding = 0;
  int have_to_start_as_daemon = 1;
  int detectable_ar = 0;
  int poll_rc;
  char *display_name = NULL;
  char *geom = NULL;
  Display *d;

  struct poptOption options_table[] = {
      {"version", 'V', 0, NULL, 'V', "prints version and exit", NULL},

      {"display", 'X', POPT_ARG_STRING, &display_name, 0,
       "Set X display to use", NULL},

      {"file", 'f', POPT_ARG_STRING, &rc_guile_file, 0,
       "Use an alternative rc file", NULL},

      {"geometry", 'g', POPT_ARG_STRING, &geom, 0,
       "size and position of window open with -k or-m option", NULL},

      {"verbose", 'v', 0, &verbose, 'v', "be verbose for debugging purposes",
       NULL},

      {"poll_rc", 'p', 0, &poll_rc, 0, "Poll the config for updates", NULL},

      {"show", 's', 0, &have_to_show_binding, 0, "Show the actual keybindings",
       NULL},

      {"key", 'k', 0, &have_to_get_binding, 0, "Identify one keypress", NULL},

      {"multikey", 'm', 0, NULL, 'm', "Identify multiple keypresses", NULL},

      {"nodaemon", 'n', 0, NULL, 'n', "don't start as daemon", NULL},

      {"detectable-ar", 'a', 0, &detectable_ar, 'a',
       "something something autorepeat", NULL},

      POPT_AUTOHELP{NULL, 0, 0, NULL, 0, NULL, NULL}};

  poptContext opt_con = poptGetContext(NULL, ARGC, argv, options_table, 0);
  poptSetOtherOptionHelp(opt_con, "[OPTIONS]*");

  while ((c = poptGetNextOpt(opt_con)) >= 0) {
    switch (c) {
    case 'V':
      fprintf(stderr, "xbindkeys %s by Philippe Brochard & @Hudmont\n",
              PACKAGE_VERSION);
      exit(1);
      break;
    case 'v':
    case 'n':
      have_to_start_as_daemon = 0;
      break;
    case 'm':
      have_to_get_binding = 2;
      break;
    }
  }

  if (rc_guile_file == NULL) {
    rc_guile_file = default_file;
  }

  if (!rc_file_exist(rc_guile_file)) {
    exit(-1);
  }

  if (have_to_start_as_daemon && !have_to_show_binding && !have_to_get_binding) {
    start_as_daemon();
  }

  if (!display_name) {
    display_name = XDisplayName(NULL);
  }

  if (verbose) {
    printf("displayName = %s\n", display_name);
    printf("rc guile file = %s\n", rc_guile_file);
  }

  d = start(display_name);
  // current_display = d;

  if (detectable_ar) {
    Bool supported_rtrn;
    XkbSetDetectableAutoRepeat(d, True, &supported_rtrn);

    if (!supported_rtrn) {
      fprintf(stderr, "Could not set detectable autorepeat\n");
    }
    
  }
  #ifdef DEBUG
  printf("Autorepeat brach resolved!\n");
  #endif
  get_offending_modifiers(d);
  #ifdef DEBUG
  printf("Offending modifiers retrieved!\n");
  #endif
  if (have_to_get_binding) {
    get_key_binding(d, have_to_get_binding, geom);
    end_it_all(d);
    exit(0);
  }
  struct passed_data to_pass = {.rc_guile_file = rc_guile_file,
                                .have_to_show_binding = have_to_show_binding,
                                .poll_rc = poll_rc,
                                .d = d,
                                .verbose = verbose};
  //free(home);

#ifdef DEBUG
  printf("Starting in guile mode...\n");
#endif

  scm_boot_guile(0, (char **)NULL, inner_main, (void *)(&to_pass));

  return 0; /* not reached ...*/
}

/*void handle(Display *d, XEvent *e, int verbose)
{
        print_key (d, &keys[i], verbose);
        adjust_display(&e.xany);
        start_command_key (&keys[i]);
}
*/

void init_finalize(Display *d, char *rc_guile_file, int have_to_show_binding,
                   int verbose) {
  // Config loading needs to be done in guilemode, therefore it's here
  if (get_rc_guile_file(d, rc_guile_file, verbose) != 0) {
    exit(-1);
  }
  // this option requires the preloaded conf file
  if (have_to_show_binding) {
    show_key_binding(d, verbose);
    end_it_all(d);
    exit(0);
  }

  grab_keys(d, verbose);

  /* This: for restarting reading the RC file if got a HUP signal */
  signal(SIGHUP, catch_hup_signal);
  /* and for reaping dead children */
  signal(SIGCHLD, catch_chld_signal);
}

void inner_main(void *passed_data, int argc, char **argv) {
  struct passed_data *params = (struct passed_data *)passed_data;
  struct passed_data p = *params;

  init_finalize(p.d, p.rc_guile_file, p.have_to_show_binding, p.verbose);

#ifdef DEBUG
  printf("starting loop...\n");
#endif

  XEvent e;

  time_t rc_guile_file_changed = 0;
  struct stat rc_guile_file_info;

  XSetErrorHandler((XErrorHandler)&null_x_error);

  if (p.poll_rc) {

    stat(p.rc_guile_file, &rc_guile_file_info);
    rc_guile_file_changed = rc_guile_file_info.st_mtime;
  }

  while (True) {
    while ((got_hup || p.poll_rc) && !XPending(p.d)) {

      // if the rc guile file has been modified, reload it
      stat(p.rc_guile_file, &rc_guile_file_info);

      if (rc_guile_file_info.st_mtime != rc_guile_file_changed) {
        reload_rc_file(p.d, p.rc_guile_file, p.verbose);
#ifdef DEBUG
        printf("The configuration file has been modified, reload it\n");
#endif
        rc_guile_file_changed = rc_guile_file_info.st_mtime;
        got_hup = 0;
      }

      usleep(SLEEP_TIME * 1000);
    }

    XNextEvent(p.d, &e);

    switch (e.type) {
    case KeyPress:
#ifdef DEBUG
      printf("Key press !\n");
      printf("e.xkey.keycode=%d\n", e.xkey.keycode);
      printf("e.xkey.state=%d\n", e.xkey.state);
#endif

      e.xkey.state &= ~(numlock_mask | capslock_mask | scrolllock_mask);

      for (int i = 0; i < nb_keys; i++) {
        if (keys[i].type == SYM && keys[i].event_type == PRESS) {
          if (e.xkey.keycode == XKeysymToKeycode(p.d, keys[i].key.sym) &&
              e.xkey.state == keys[i].modifier) {
            print_key(p.d, &keys[i], p.verbose);
            adjust_display(&e.xany);
            start_command_key(&keys[i]);
          }
        } else if (keys[i].type == CODE && keys[i].event_type == PRESS) {
          if (e.xkey.keycode == keys[i].key.code &&
              e.xkey.state == keys[i].modifier) {
            print_key(p.d, &keys[i], p.verbose);
            adjust_display(&e.xany);
            start_command_key(&keys[i]);
          }
        }
      }
      break;

    case KeyRelease:
#ifdef DEBUG
      printf("Key release !\n");
      printf("e.xkey.keycode=%d\n", e.xkey.keycode);
      printf("e.xkey.state=%d\n", e.xkey.state);
#endif

      e.xkey.state &= ~(numlock_mask | capslock_mask | scrolllock_mask);

      for (int i = 0; i < nb_keys; i++) {
        if (keys[i].type == SYM && keys[i].event_type == RELEASE) {
          if (e.xkey.keycode == XKeysymToKeycode(p.d, keys[i].key.sym) &&
              e.xkey.state == keys[i].modifier) {
            print_key(p.d, &keys[i], p.verbose);
            adjust_display(&e.xany);
            start_command_key(&keys[i]);
          }
        } else if (keys[i].type == CODE && keys[i].event_type == RELEASE) {
          if (e.xkey.keycode == keys[i].key.code &&
              e.xkey.state == keys[i].modifier) {
            print_key(p.d, &keys[i], p.verbose);
            adjust_display(&e.xany);
            start_command_key(&keys[i]);
          }
        }
      }
      break;

    case ButtonPress:
#ifdef DEBUG
      printf("Button press !\n");
      printf("e.xbutton.button=%d\n", e.xbutton.button);
      printf("e.xbutton.state=%d\n", e.xbutton.state);
#endif

      e.xbutton.state &=
          0x1FFF &
          ~(numlock_mask | capslock_mask | scrolllock_mask | Button1Mask |
            Button2Mask | Button3Mask | Button4Mask | Button5Mask);

      for (int i = 0; i < nb_keys; i++) {
        if (keys[i].type == BUTTON && keys[i].event_type == PRESS) {
          if (e.xbutton.button == keys[i].key.button &&
              e.xbutton.state == keys[i].modifier) {
            // printf("Replay!!!\n");
            // ungrab_all_keys(p.d);
            // XPutBackEvent(p.d, &e);
            // sleep(1);
            // grab_keys(p.d, p.verbose);
            print_key(p.d, &keys[i], p.verbose);
            adjust_display(&e.xany);
            start_command_key(&keys[i]);
          }
        }
      }
      break;

    case ButtonRelease:
#ifdef DEBUG
      printf("Button release !\n");
      printf("e.xbutton.button=%d\n", e.xbutton.button);
      printf("e.xbutton.state=%d\n", e.xbutton.state);
#endif

      e.xbutton.state &=
          0x1FFF &
          ~(numlock_mask | capslock_mask | scrolllock_mask | Button1Mask |
            Button2Mask | Button3Mask | Button4Mask | Button5Mask);

      for (int i = 0; i < nb_keys; i++) {
        if (keys[i].type == BUTTON && keys[i].event_type == RELEASE) {
          if (e.xbutton.button == keys[i].key.button &&
              e.xbutton.state == keys[i].modifier) {
            print_key(p.d, &keys[i], p.verbose);
            adjust_display(&e.xany);
            start_command_key(&keys[i]);
          }
        }
      }
      break;

    default:
      break;
    }
  } /*  infinite loop */

#ifdef DEBUG
  printf("ending...\n");
#endif
  end_it_all(p.d);

#ifdef AVOID_KNOWN_HARMLESS_WARNINGS
  argc = argc;
  argv = argv;
#endif
}
